src/OrbitGl/TrackContainer.cpp

```diff
@@ -0,0 +1,515 @@
+                                                          std::optional<uint32_t> thread_id) const {
+  const TimerInfo* previous_timer = nullptr;
+  uint64_t goal_time = std::numeric_limits<uint64_t>::lowest();
+  std::vector<const TimerChain*> chains = GetAllThreadTrackTimerChains();
+  for (const TimerChain* chain : chains) {
+    for (const auto& block : *chain) {
+      if (!block.Intersects(goal_time, current_time)) continue;
+      for (uint64_t i = 0; i < block.size(); i++) {
+        const TimerInfo& timer_info = block[i];
+        auto timer_end_time = timer_info.end();
+        if ((timer_info.function_id() == function_address) &&
+            (!thread_id || thread_id.value() == timer_info.thread_id()) &&
+            (timer_end_time < current_time) && (goal_time < timer_end_time)) {
+          previous_timer = &timer_info;
+          goal_time = timer_end_time;
+        }
+      }
+    }
+  }
+  return previous_timer;
+}
+
+const TimerInfo* TrackContainer::FindNextFunctionCall(uint64_t function_address,
+                                                      uint64_t current_time,
+                                                      std::optional<uint32_t> thread_id) const {
+  const TimerInfo* next_timer = nullptr;
+  uint64_t goal_time = std::numeric_limits<uint64_t>::max();
+  std::vector<const TimerChain*> chains = GetAllThreadTrackTimerChains();
+  for (const TimerChain* chain : chains) {
+    ORBIT_CHECK(chain != nullptr);
+    for (const auto& block : *chain) {
+      if (!block.Intersects(current_time, goal_time)) continue;
+      for (uint64_t i = 0; i < block.size(); i++) {
+        const TimerInfo& timer_info = block[i];
+        auto timer_end_time = timer_info.end();
+        if ((timer_info.function_id() == function_address) &&
+            (!thread_id || thread_id.value() == timer_info.thread_id()) &&
+            (timer_end_time > current_time) && (goal_time > timer_end_time)) {
+          next_timer = &timer_info;
+          goal_time = timer_end_time;
+        }
+      }
+    }
+  }
+  return next_timer;
+}
+
+std::vector<const TimerInfo*> TrackContainer::GetAllTimersForHookedFunction(
+    uint64_t function_address) const {
+  std::vector<const TimerInfo*> timers;
+  std::vector<const TimerChain*> chains = GetAllThreadTrackTimerChains();
+  for (const TimerChain* chain : chains) {
+    ORBIT_CHECK(chain != nullptr);
+    for (const auto& block : *chain) {
+      for (uint64_t i = 0; i < block.size(); i++) {
+        const TimerInfo& timer = block[i];
+        if (timer.function_id() == function_address) timers.push_back(&timer);
+      }
+    }
+  }
+  return timers;
+}
+
+void TrackContainer::DoUpdateLayout() {
+  track_manager_->UpdateTrackListForRendering();
+  UpdateTracksPosition();
+
+  // This is called to make sure the current scrolling value is correctly clamped
+  // in case any changes in track visibility occured before
+  SetVerticalScrollingOffset(vertical_scrolling_offset_);
```

Comment:
To be sure, we've had this before in `TimeGraph::DoUpdateLayout`, right?

Reply:
Exactly, copy pasted from there.